---
title: "Analyses"
output:
  github_document:
    toc: yes
    toc_depth: 2
---

# DADA2

## Préparation des données

```{r}
miseq_path <- "./DataAll" # CHANGE to the directory containing the fastq files after unzipping.
list.files(miseq_path)
```

```{r}
# Forward and reverse fastq filenames have format: SAMPLENAME_1.fastq and SAMPLENAME_2.fastq
fnFs <- sort(list.files(miseq_path, pattern="_1.fastq", full.names = TRUE))
fnRs <- sort(list.files(miseq_path, pattern="_2.fastq", full.names = TRUE))
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)
```

## Inspection du profil de qualité des reads

Lorsque que l’on séquence avec Illumina, à chaque base ajoutée, on va y associer un score de qualité ou Qscore (par exemple la chance d’avoir un C si la base est un C) qui va évoluer au cours du séquencage.

Les fowards:
```{r}
plotQualityProfile(fnFs[1:2])
```

Les reverses:
```{r}
plotQualityProfile(fnRs[1:2])
```

## Filtrer et couper
```{r}
filt_path <- file.path(miseq_path, "filtered") # Place filtered files in filtered/ subdirectory
if(!file_test("-d", filt_path)) dir.create(filt_path)
filtFs <- file.path(filt_path, "filtered", paste0(sample.names, "_Ffilt.fastq.gz"))
filtRs <- file.path(filt_path, "filtered", paste0(sample.names, "_Rfilt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names
```

```{r}
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(250,200), trimLeft =8,
              maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE)
head(out)
```

## Taux d'erreur

```{r}
errF <- learnErrors(filtFs, multithread=TRUE)
```

```{r}
errR <- learnErrors(filtRs, multithread=TRUE)
```

```{r}
plotErrors(errF, nominalQ=TRUE)
```

## Interférences des échantillons

```{r}
dadaFs <- dada(filtFs, err=errF, multithread=TRUE)
```

```{r}
dadaRs <- dada(filtRs, err=errR, multithread=TRUE)
```

## Appariment des reads complémentaires

On va apparier les fowards avec leur reverses complementaires. Si nous n’avons pas de correspondence parfaite, ils seront retirés des donnees.


```{r}
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=TRUE)
```


```{r}
# Inspect the merger data.frame from the first sample
head(mergers[[1]])
```

## Construction de la table de sequence

On prépare la table:
```{r}
seqtab <- makeSequenceTable(mergers)
dim(seqtab)
```

```{r}
# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))
```

La premiere ligne donne la taille des sequences, la deuxieme donne le nombre de sequence presentant cette taille.

## Supression des chimeres

```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
```
```{r}
dim(seqtab.nochim)
```

```{r}
sum(seqtab.nochim)/sum(seqtab)
```

## Suivie de l’affinement de notre set de donnees final

On va suivre a chaque etape le nombre de sequence qui n’ont pas ete filtre et donc retire de notre set de donnees. On fait un suivi du traitement des données.

```{r}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
head(track)
```
```{bash}
wget https://zenodo.org/record/3986799/files/silva_nr99_v138_train_set.fa.gz?download=1
```

## Assignation de la taxonomie

```{r}
taxa <- assignTaxonomy(seqtab.nochim,"~/silva_nr99_v138_train_set.fa.gz?download=1", multithread = TRUE)
```


remarque : le code suivant est a faire tourner si le temps. il n'est pa encore fait actuellement 
taxa <- addSpecies(taxa, "~/silva_species_assignment_v138.fa.gz?download=1")


```{r}
taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)
```
 
 
# Analyses phyloseq

On prépare l'objet phyloseq

```{r}
library(ggplot2)
theme_set(theme_bw())
```

Remarque: pour la suite, il faut que je reussis a ne selectionner que quelques eponges de la liste (celle avec plusieurs echantillons) pour que les graphes soient plus parlant 

Mais pour l'instant, je n'ai pas reussi .... 

```{r}
samples.out <- rownames(seqtab.nochim)
subject <- samples.out
echantillon <- substr(subject,0,26)
subject <- substr(subject,0,26)
samples.out2 <- samples.out
samples.out2[c(1, 6, 7, 9, 15, 16, 17, 21, 22, 23)] <- 09.2015
samples.out2[c(8, 10, 12, 14, 18, 25)] <- 08.2015
samples.out2[c(2, 3, 4, 5, 11, 13, 19, 20, 24, 26, 27)] <- 05.2015
samples.out3 <- samples.out
samples.out3[c(2, 3, 4, 5, 11, 13, 19, 20, 24, 26, 27)] <- "Viet Nam: Lang Co Bay"
samples.out3[c(1, 6, 7, 9, 15, 16, 17, 21, 22, 23)] <- "Viet Nam: Hon mun island"
samples.out3[c(8, 10, 12, 14, 18, 25)] <- "Viet Nam: Con Co island"
samples.out4 <- samples.out
samples.out4[c(1)] <- "Amphimedon sp."
samples.out4[c(2)] <- "Cinachyrella schulzei"
samples.out4[c(3, 4)] <- "Haliclona amboinensis"
samples.out4[c(5)] <- "Haliclona fascigera"
samples.out4[c(6, 14)] <- "Clathria reinwardti"
samples.out4[c(7, 16, 17, 19)] <- "Axinyssa topsenti"
samples.out4[c(8, 20, 27)] <- "Xestospongia testudinaria"
samples.out4[c(9)] <- "Haplosclerida sp."
samples.out4[c(10, 15)] <- "Spirastrella sp."
samples.out4[c(11, 21)] <- "Dactylospongia sp."
samples.out4[c(12)] <- "Spheciospongia vesparium"
samples.out4[c(13)] <- "Amphimedon compressa"
samples.out4[c(18)] <- "Halichondria sp."
samples.out4[c(22)] <- "Tedania sp."
samples.out4[c(23)] <- "Niphatidae sp."
samples.out4[c(24)] <- "Rhabdastrella globostellata"
samples.out4[c(25)] <- "Axos cliftoni"
samples.out4[c(26)] <- "Terpios aploos"
environnement <- samples.out3
date <- as.integer(samples.out2)
hote <- samples.out4
samdf <- data.frame(Subject=subject, Echantillon=echantillon, Date=date, Environnement=environnement, Hote=hote)
samdf$When <- "2015"
samdf$When[samdf$Year == "05.2015"] <- "Mai2015"
samdf$When[samdf$Year == "08.2015"] <- "Aout2015"
samdf$When[samdf$Year == "09.2015"] <- "Septembre2015"
rownames(samdf) <- samples.out
```


code suivant pour essayer de ne sélectionner que quelques lignes du tableua et donc ne pas utiliser les éponges prélevés une seule fois , en cours d'élaboration :

samples.out <- rownames(seqtab.nochim)
subject2 <- samples.out
echantillon2 <- substr(subject2, 0, 27)
subject2 <- substr(subject2, 0, 27)
samples.out22 <- samples.out
samples.out22[c(6, 7, 15, 16, 17, 21)] <- 09.2015
samples.out22[c(8, 10, 14)] <- 08.2015
samples.out22[c(3, 4, 11, 19, 20, 27)] <- 05.2015
samples.out32 <- samples.out
samples.out32[c(3, 4, 11, 19, 20)] <- "Viet Nam: Lang Co Bay"
samples.out32[c(6, 7, 15, 16, 17, 21)] <- "Viet Nam: Hon mun island"
samples.out32[c(8, 10, 14, 27)] <- "Viet Nam: Con Co island"
samples.out42 <- samples.out
samples.out42[c(3, 4)] <- "Haliclona amboinensis"
samples.out42[c(6, 14)] <- "Clathria reinwardti"
samples.out42[c(7, 16, 17, 19)] <- "Axinyssa topsenti"
samples.out42[c(8, 20, 27)] <- "Xestospongia testudinaria"
samples.out42[c(10, 15)] <- "Spirastrella sp."
samples.out42[c(11, 21)] <- "Dactylospongia sp."
environnement2 <- samples.out32
date2 <- as.integer(samples.out22)
hote2 <- samples.out42
samdf2 <- data.frame(Subject=subject2, Echantillon=echantillon2, Date=date2, Environnement=environnement2, Hote=hote2)
samdf2$When <- "2015"
samdf2$When[samdf$Year == "05.2015"] <- "Mai2015"
samdf2$When[samdf$Year == "08.2015"] <- "Aout2015"
samdf2$When[samdf$Year == "09.2015"] <- "Septembre2015"
rownames(samdf2) <- samples.out


```{r}
library(phyloseq)
ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               sample_data(samdf), 
               tax_table(taxa))
ps <- prune_samples(sample_names(ps) != "Mock", ps) # Remove mock sample
```
 
```{r}
dna <- Biostrings::DNAStringSet(taxa_names(ps))
names(dna) <- taxa_names(ps)
ps <- merge_phyloseq(ps, dna)
taxa_names(ps) <- paste0("ASV", seq(ntaxa(ps)))
ps
```

## Alpha diversité

On va calculer des indices d’alpha diversite: Shannon et l'inverse de Simpson.


```{r}
plot_richness(ps, x="Echantillon", measures=c("Shannon"), color="hote")
```
```{r}
plot_richness(ps, x="environnement", measures=c("Shannon"), color="Hote")
```

 
```{r}
plot_richness(ps, x="Echantillon", measures=c("InvSimpson"), color="Hote")
```

```{r}
plot_richness(ps, x="environnement", measures=c("InvSimpson"), color="Hote")
```

## Heatmap

```{r}
library(ggplot2)
library(Heatplus)
library(vegan)
library(RColorBrewer)
```

```{r}
# preparation de la palette de couleur
scaleyellowred <- colorRampPalette(c("lightyellow", "red"), space = "rgb")(100)
```

```{r}
dim(seqtab.nochim)
```

```{r}
row.names(seqtab.nochim) <- seqtab.nochim$sample
seqtab.nochim <- seqtab.nochim[, -1]
```

 

## Beta-diversite

On va maitenant faire une NMDS (indice de beta diversite). Les points vont êtres places les uns en fonction des autres. On ajoute donc des relations.

```{r}
pslog <- transform_sample_counts(ps, function(x) log(1+x))
out.wuf.log <- ordinate(pslog, method = "NMDS", distance = "bray")
```
 
```{r}
library(vegan)
evals <- out.wuf.log$values$Eigenvalues
plot_ordination(pslog, out.wuf.log,  color = "Hote", shape="Environnement") +
labs(col = "Hote",shape= "Environnement")
ordiellipse(out.wuf.log, groups = evals, draw = "polygon", lty = 1, col = "grey90")

```
Non-metric multidimensional scaling (NMDS) plot derived from Bray–Curtis distances of sponge prokaryotic communities at OTUs level. The samples of the same species were grouped with ordination ellipse using function ordiellipse of vegan package

